### PR TITLE
add 4.12.0 multicore compiler variants

### PR DESCRIPTION
**Pull request by avsm**
These add two new base packages, base-domains (for indicating
the presence of Domain based parallelism) and base-effects
(for indicating the experimental effect system, which has extra
keywords that break some ecosystem packages)

This PR also introduces domainslib.0.3.0 as a library using
base-domains to illustrate how this works.

When this PR is merged, I will add followup packages that
uses these multicore compilers.

Package source: https://github.com/ocaml-multicore/multicore-opam

### CODE REVIEW

File: packages/base-domains/base-domains.base/opam [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+depends: [
+  "ocaml" {>="4.06.1"}
+  "ocaml-variants" {
+    = "4.12.0+domains+effects" |
+    = "4.12.0+domains" |
+    = "4.10.0+multicore" |
+    = "4.10.0+multicore+no-effect-syntax" |
+    = "4.06.1+multicore" |
+    = "4.04.2+multicore" |
+    = "4.02.2+multicore"
```
Comment (avsm) [inline packages/base-domains/base-domains.base/opam:15]: I'll import these more historical compilers later; this PR only includes 4.12.0*

File: packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    #"PARTIALLD=ld -r -melf_i386" {ocaml-option-32bit:installed & os="linux"}
+    # 32bit options above commented out just to reduce diff with ocaml-variants.4.12.0+options
+    "LIBS=-static" {ocaml-option-static:installed}
+    "--disable-warn-error"
+  ]
+  [make "-j%{jobs}%"]
+]
+install: [make "install"]
+url {
+  src: "git+https://github.com/ocaml-multicore/ocaml-multicore.git#4.12+domains+effects"
```
Comment (dra27) [inline packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam:49]: The use of a Git URL here is tripping opam-dune-lint... this rings a bell, but I can't remember the full story. @kit-ty-kate?

**Review by dra27 — COMMENTED** [review]
4 cosmetic deletions - the lint failure needs checking

File: packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+}
+conflicts: [
+  "ocaml-options-vanilla"   # This is not vanilla OCaml!
```
[the remark this answers is not in the document]
Comment (avsm) [reply packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam:52]: I see, so this vanilla package should only be used to install ocaml-base-compiler, then? I don't entirely understand its purpose for existence.

File: packages/base-domains/base-domains.base/opam [hunk 4]
```diff
@@ -0,0 +1,17 @@
+opam-version: "2.0"
+maintainer: "https://github.com/ocaml-multicore/multicore-opam/issues"
+description: """
+Domains-based parallelism distributed with the Multicore OCaml compiler"
+"""
+depends: [
+  "ocaml" {>="4.06.1"}
```
[the remark this answers is not in the document]
Comment (avsm) [reply packages/base-domains/base-domains.base/opam:7]: This is necessary for future PRs when I will import historical versions of the multicore compilers (for archival)

File: packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam (hunk 3 above)
[the remark this answers is not in the document]
Comment (dra27) [reply packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam:52]: No, it's purpose is to ensure that you don't get any other configuration options - i.e. flambda cannot be switched on, nnp cannot be switched, etc. It has nothing to do with the sources of the compiler hence - ocaml-**options**-vanilla!

File: packages/base-domains/base-domains.base/opam (hunk 4 above)
[the remark this answers is not in the document]
Comment (dra27) [reply packages/base-domains/base-domains.base/opam:7]: I don't follow - having `"ocaml" {>= "4.06.1"}` _prevents_ such a future PR from ever working?

File: packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam (hunk 3 above)
[the remark this answers is not in the document]
Comment (avsm) [reply packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam]: fixed in 283b0ba689c49edeac18a8223b4f1df2330c16a4

File: packages/base-effects/base-effects.base/opam [hunk 5]
```diff
@@ -0,0 +1,16 @@
+opam-version: "2.0"
+maintainer: "https://github.com/ocaml-multicore/multicore-opam/issues"
+description: """
+Effects-based variant distributed with the Multicore OCaml compiler"
+"""
+depends: [
+ "ocaml" {>= "4.12.0"}
```
[the remark this answers is not in the document]
Comment (avsm) [reply packages/base-effects/base-effects.base/opam:7]: fixed in 2c01911dec11ced6f7186d8a925fd9e76fe03e25

File: packages/base-domains/base-domains.base/opam (hunk 4 above)
[the remark this answers is not in the document]
Comment (avsm) [reply packages/base-domains/base-domains.base/opam]: fixed in 283b0ba689c49edeac18a8223b4f1df2330c16a4

File: packages/base-domains/base-domains.base/opam [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+Domains-based parallelism distributed with the Multicore OCaml compiler"
+"""
+depends: [
+  "ocaml" {>="4.06.0"}
+  "ocaml-variants" {
+    = "4.12.0+domains+effects" |
+    = "4.12.0+domains" |
+    = "4.10.0+multicore" |
+    = "4.10.0+multicore+no-effect-syntax" |
+    = "4.06.1+multicore" |
```
Comment (dra27) [inline packages/base-domains/base-domains.base/opam:13]:
```suggestion
    = "4.06.1+multicore"
```

File: packages/base-effects/base-effects.base/opam [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+ "ocaml-variants" {
+    = "4.12.0+domains+effects" |
```
Comment (dra27) [inline packages/base-effects/base-effects.base/opam:9]:
```suggestion
    = "4.12.0+domains+effects"
```

File: packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam (hunk 2 above)
Comment (avsm) [reply packages/ocaml-variants/ocaml-variants.4.12.0+domains+effects/opam:49]: thanks!

**Comment by kit-ty-kate** [comment]
```
#=== ERROR while compiling ocaml-variants.4.12.0+domains+effects ==============#
# context              2.1.0~rc2 | linux/x86_64 | ocaml-variants.4.13.0+trunk | pinned(git+https://github.com/ocaml-multicore/ocaml-multicore.git#4.12+domains+effects#2f7fea60)
# path                 ~/.opam/4.13/.opam-switch/build/ocaml-variants.4.12.0+domains+effects
# command              ~/.opam/opam-init/hooks/sandbox.sh build make -j47
# exit-code            2
# env-file             ~/.opam/log/ocaml-variants-85-6dfafa.env
# output-file          ~/.opam/log/ocaml-variants-85-6dfafa.out
### output ###
# ./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -I runtime -c middle_end/flambda/simplify_primitives.ml
# ../yacc/ocamlyacc -v parser.mly
# ../yacc/ocamlyacc -v parser.mly
# ./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -I runtime -c middle_end/flambda/inlining_stats.ml
# ../boot/ocamlrun ../boot/ocamllex -q lexer.mll
# ../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot -use-prims ../runtime/primitives -c -absname -w +a-4-9-41-42-44-45-48 -warn-error A -safe-string -strict-sequence -strict-formats -bin-annot table.mli
# ../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot -use-prims ../runtime/primitives -c -absname -w +a-4-9-41-42-44-45-48 -warn-error A -safe-string -strict-sequence -strict-formats -bin-annot cset.ml
# ../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot -use-prims ../runtime/primitives -c -absname -w +a-4-9-41-42-44-45-48 -warn-error A -safe-string -strict-sequence -strict-formats -bin-annot syntax.mli
# ../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot -use-prims ../runtime/primitives -c -absname -w +a-4-9-41-42-44-45-48 -warn-error A -safe-string -strict-sequence -strict-formats -bin-annot table.ml
# ../boot/ocamlrun ../ocamlc -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -nostdlib -principal -safe-string -strict-formats  -nopervasives -c camlinternalFormatBasics.ml
# ../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot -use-prims ../runtime/primitives -c -absname -w +a-4-9-41-42-44-45-48 -warn-error A -safe-string -strict-sequence -strict-formats -bin-annot syntax.ml
# ../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot -use-prims ../runtime/primitives -c -absname -w +a-4-9-41-42-44-45-48 -warn-error A -safe-string -strict-sequence -strict-formats -bin-annot parser.mli
# ../boot/ocamlrun ../ocamlc -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -nostdlib -principal -safe-string -strict-formats  -nopervasives -no-alias-deps -w -49  -pp "$AWK -f ./expand_module_aliases.awk" -c stdlib.mli
# ../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot -use-prims ../runtime/primitives -c -absname -w +a-4-9-41-42-44-45-48 -warn-error A -safe-string -strict-sequence -strict-formats -bin-annot lexgen.mli
# ../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot -use-prims ../runtime/primitives -c -absname -w +a-4-9-41-42-44-45-48 -warn-error A -safe-string -strict-sequence -strict-formats -bin-annot lexer.mli
# ./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -I runtime -c middle_end/flambda/inlining_decision_intf.mli
# ./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -I runtime -c middle_end/flambda/inline_and_simplify_aux.ml
# File "/home/opam/.opam/4.13/.opam-switch/build/ocaml-variants.4.12.0+domains+effects/lex/lexer.mli", line 16, characters 27-39:
# 16 | val main: Lexing.lexbuf -> Parser.token
#                                 ^^^^^^^^^^^^
# Error: Unbound type constructor Parser.token
# ./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -I runtime -c middle_end/flambda/extract_projections.mli
# make[5]: *** [Makefile:81: lexer.cmi] Error 2
# make[5]: *** Waiting for unfinished jobs....
```
It looks like +domains+effects has some parallel compilation issues ^^"" We might want to use -j1 here

**Comment by dra27** [comment]
Actually, this is an instance of https://github.com/ocaml/opam-repository/issues/18726.

**Comment by dra27** [comment]
(cf. the two parallel calls to `../yacc/ocamlyacc -v parser.mly`)

**Comment by kit-ty-kate** [comment]
Fair enough, looks good then. Thanks!

**Pull request merged**